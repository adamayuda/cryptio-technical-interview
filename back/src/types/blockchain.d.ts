import { ICryptoTransactionsHistorical } from 'src/types/crypto';

export interface IBlockchainTransactionOut {
  value: string | number;
  hash?: string;
  script?: string;
  addr?: string;
}

export interface IBlockchainTransactionInput {
  prev_out: {
    hash?: string;
    value: string | number;
    addr?: string;
    tx_index?: string;
    n?: string;
  };
  script?: string;
}

export interface IBlockchainTransaction {
  hash: string;
  ver: number;
  vin_sz: number;
  ICryptoTransactionsHistorical;
  vout_sz: number;
  time: Date;
  lock_time: string;
  size: number;
  relayed_by: string;
  block_height: number;
  tx_index: string;
  inputs: IBlockchainTransactionInput[];
  out: IBlockchainTransactionOut[];
}

export interface IBlockchainAddress {
  hash160: string;
  address: string;
  n_tx: number;
  n_unredeemed: number;
  total_received: number;
  total_sent: number;
  final_balance: number;
  txs: IBlockchainTransaction[];
}
