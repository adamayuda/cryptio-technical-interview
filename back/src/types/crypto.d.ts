export interface ICryptoTransactionsMatrix {
  amount: number;
  time: Date;
}

export interface ICryptoService {
  getHistoricalBalances({
    address,
  }: {
    address: string;
  }): Promise<ICryptoTransactionsMatrix[]>;
}
