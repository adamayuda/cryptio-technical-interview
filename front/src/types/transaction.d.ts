export interface ITransactionsMatrix {
  balance: number;
  transactions: ITransaction[];
}

export interface ITransaction {
  hash: string;
  amount: number;
  time: number;
}
