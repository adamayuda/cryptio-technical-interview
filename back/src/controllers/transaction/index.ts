import { Request, Response } from 'express';

export interface ITransactionController {
  health(req: Request, res: Response): Promise<Response>;
}

export default class TransactionController implements ITransactionController {
  health = async (req: Request, res: Response): Promise<Response> => {
    return res.status(200).json({});
  };
}
