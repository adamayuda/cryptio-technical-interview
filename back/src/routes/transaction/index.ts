import { Router } from 'express';
import TransactionController from '../../controllers/transaction';

const transactionController = new TransactionController();

const route = Router();

export default (app: Router) => {
  app.use('/transactions', route);

  route.post('/health', transactionController.health);
};
