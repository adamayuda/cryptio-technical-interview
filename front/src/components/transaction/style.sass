@import "src/styles/colors.sass"
@import "src/styles/device.sass"

.transaction
  margin-top: 10px
  background-color: $color-main
  padding: 10px 20px
  border: 1px solid $color-third
  border-radius: 3px
  color: $color-third

  .header
    display: grid
    grid-template-columns: auto 100px

    h3
      margin: 0

    .amount
      border-radius: 3px
      padding: 5px 10px
      text-align: right

      &.green
        background-color: $color-green-light
        color: $color-green
      &.red
        background-color: $color-red-light
        color: $color-red
  .date
    text-align: right
    margin-top: 10px

