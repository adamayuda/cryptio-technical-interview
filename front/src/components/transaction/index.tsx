import "src/components/transaction/style.sass";

import React, { FC } from "react";
import { ITransaction } from "src/types/transaction";
import Moment from "react-moment";
import PropTypes from "prop-types";

interface ITransactionProps {
  transaction: ITransaction;
}

export const Transaction: FC<ITransactionProps> = ({ transaction }) => {
  return (
    <div className="transaction">
      <div className="header">
        <h3>{transaction.hash}</h3>
        <div className={`amount ${transaction.amount < 0 ? "red" : "green"}`}>
          {transaction.amount * 0.00000001}
        </div>
      </div>
      <div className="date">
        <Moment format="YYYY/MM/DD hh:mm">{transaction.time * 1000}</Moment>
      </div>
    </div>
  );
};

Transaction.propTypes = {
  transaction: PropTypes.any,
};
