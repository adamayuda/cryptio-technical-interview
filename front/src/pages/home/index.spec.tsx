import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { ITransaction } from "src/types/transaction";
import { Home } from ".";
import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import axios from "axios";
import { render } from "@testing-library/react";

configure({ adapter: new Adapter() });
describe("src/pages/home/index.tsx", () => {
  it("Render Home page", async () => {
    const renderedHome = render(
      <Router>
        <Home />
      </Router>,
    );

    expect(renderedHome.container).toMatchSnapshot();
  });

  it("should pass a selected value to the onChange function", () => {
    const component = shallow(<Home />);
    component
      .find("input")
      .at(0)
      .simulate("change", {
        target: {
          value: "Change function",
        },
      });

    expect(component).toMatchSnapshot();
  });

  it("should add an item based on the value in the state", () => {
    const component = shallow(<Home />);
    const preventDefault = jest.fn();
    // component.setState({
    //   address: "ok",
    // });
    axios.get = jest.fn().mockResolvedValue({
      data: {
        balance: 0,
        transactions: [
          {
            hash: "hash",
            amount: 1,
            time: 123,
          },
          {
            hash: "hash",
            amount: -1,
            time: 123,
          },
        ],
      },
    });
    component.find("form").simulate("submit", { preventDefault });
    expect(component).toMatchSnapshot();
    expect(preventDefault).toBeCalled();
  });
});
