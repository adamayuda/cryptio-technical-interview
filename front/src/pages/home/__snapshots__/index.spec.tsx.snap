// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/home/index.tsx Render Home page 1`] = `
<div>
  <div
    class="home"
  >
    <div
      class="container"
    >
      <form>
        <input
          placeholder="Address"
          type="text"
          value=""
        />
        <input
          class="submit"
          type="submit"
          value="Search"
        />
      </form>
    </div>
  </div>
</div>
`;

exports[`src/pages/home/index.tsx should add an item based on the value in the state 1`] = `ShallowWrapper {}`;

exports[`src/pages/home/index.tsx should pass a selected value to the onChange function 1`] = `ShallowWrapper {}`;
